shared/chat/conversation/messages/index.js

```diff
@@ -1,116 +1,59 @@
+import Wrapper from './wrapper/container'
+import {Box} from '../../../common-adapters'
 
-import type {Options} from './index'
-
-const factory = (options: Options) => {
-  const {
-    message,
-    includeHeader,
-    key,
-    isEditing,
-    isFirstNewMessage,
-    isSelected,
-    onAction,
-    onLoadAttachment,
-    onOpenConversation,
-    onOpenInFileUI,
-    onOpenInPopup,
-    onRetry,
-    onRetryAttachment,
-    onShowEditor,
-    style,
-    you,
-    metaDataMap,
-    followingMap,
-    moreToLoad,
-  } = options
-
-  if (!message) {
-    return <Box key={key} style={style} />
-  }
-
-  switch (message.type) {
-    case 'Text':
-      return <MessageText
-        key={key}
-        you={you}
-        metaDataMap={metaDataMap}
-        followingMap={followingMap}
-        style={style}
-        message={message}
-        onRetry={onRetry}
-        onShowEditor={onShowEditor}
-        includeHeader={includeHeader}
-        isFirstNewMessage={isFirstNewMessage}
+const factory = (
+  messageKey: Constants.MessageKey,
+  prevMessageKey: ?Constants.MessageKey,
+  onAction: (message: Constants.ServerMessage, event: any) => void,
+  onShowEditor: (message: Constants.ServerMessage, event: any) => void,
+  isSelected: boolean,
+  measure: () => void
+) => {
+  const kind = Constants.messageKeyKind(messageKey)
+  switch (kind) {
+    case 'header':
+      return <Header messageKey={messageKey} />
+    case 'messageIDAttachment':
+      return <Wrapper
+        innerClass={Attachment}
         isSelected={isSelected}
-        isEditing={isEditing}
+        measure={measure}
+        messageKey={messageKey}
         onAction={onAction}
-        />
-    case 'Supersedes':
-      return <ProfileResetNotice
-        onOpenOlderConversation={() => onOpenConversation(message.supersedes)}
-        username={message.username}
-        style={style}
-        key={`supersedes:${message.supersedes}`}
-        />
-    case 'Timestamp':
-      return <Timestamp
-        timestamp={formatTimeForMessages(message.timestamp)}
-        key={message.key}
-        style={style}
-        />
-    case 'Attachment':
-      return <AttachmentMessageRender
-        key={key}
-        style={style}
-        you={you}
-        metaDataMap={metaDataMap}
-        followingMap={followingMap}
-        message={message}
-        onRetry={() => onRetryAttachment(message)}
-        includeHeader={includeHeader}
-        isFirstNewMessage={isFirstNewMessage}
-        onLoadAttachment={onLoadAttachment}
-        onOpenInFileUI={onOpenInFileUI}
-        onOpenInPopup={onOpenInPopup}
-        messageID={message.messageID}
+        onShowEditor={onShowEditor}
+        prevMessageKey={prevMessageKey} />
+    case 'error': // fallthrough
+    case 'errorInvisible': // fallthrough
+    case 'messageIDError':
+      return <ErrorMessage messageKey={messageKey} />
+    case 'outboxIDText': // fallthrough
+    case 'messageIDText':
+      return <Wrapper
```

Comment:
Passing the innerclass to stop render thrashing on this thing that is used so much. if we render a child (i tried this originally) as the child changes the parent has to re-render. this way its 'one' component and more disconnected. Wrapper and the child are also connected